ejb/WFLY-6892-ejb-access-logging.adoc

```diff
@@ -0,0 +1,309 @@
+= EJB Access Logging
+:author:            Cheng Fang
+:email:             cfang@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFLY-6892 (Access logging for EJBs)
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/WFCORE-4355 (Add utility for event logging such as audit and access logging)
+* https://issues.jboss.org/browse/EAP7-523 (Access logging for EJBs)
+* https://issues.jboss.org/browse/EAP7-699 (Enhanced Audit Logging - RFC support and reliability/speed customization)
+* https://issues.jboss.org/browse/WFLY-11031 (Json encoding of Access Log Events)
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:szhantem@redhat.com[Sultan Zhantemirov]
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE. 
+// Discuss with QE during the Kickoff state to decide this
+[x] Engineering
+
+[x] QE
+
+=== Affected Projects or Components
+
+* ejb3
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+* implement the access logging feature similar to WildFly web access logging.
+
+* add access-logging definition to wildfly-ejb3 schema (see next section below)
+
+* configure ejb access logging with CLI
+
+* enable and disable ejb access logging
+
+* configure how to rotate ejb access logs if applicable
+
+** default should be OFF for performance consideration
+
+* configure output location and format
+
+** directory/file
+
+** stdout
+
+** JSON
+
+* choose from pre-defined logging formats (need to adapt for ejb invocation):
+
+** short: date time ip user ejb method
+
+*** For example: 2019-05-05 12:23:27,003 127.0.0.1 admin hello/helloBean hello
+
+** long : date time ip user ejb method invocation event
+
+*** For example: 2019-05-05 12:23:27,003 127.0.0.1 user1 hello/helloBean hello 1234 received
+                 2019-05-05 12:23:30,003 127.0.0.1 user1 hello/helloBean hello 1234 finished
+
+** default: short
+
+* for more customization, define which attributes to include in ejb access logging with the following attributes:
+
+** date:
+** time:
+** timezone:
+** ip: client ip address
+** user: authenticated caller user name
+** ejb: ejb name (may need to prefix with application name)
+** method: ejb method
+** invocation: invocation id
+** event: event type ("received", "failed", "completed"; subject to available events from ejb container)
+** host: client host name
+** port: client port number
+** protocol: protocol used for this invocation
+** thread: thread name for this invocation
+** server: local server name
+
+* For example, one may configure the following custom pattern, which can be applied to both pattern formatter or
+JSON formatter:
+
+** pattern="date time client ejb method server thread"
+
+
+=== New elements in wildfly-ejb3_6_0.xsd schema
```

Comment:
@chengfang Is it necessary to introduce so complex model addition?

Seems to me a lot of types from logging subsystem are duplicated here. Couldn't be leveraged logging subsystem through capability reference?

Or when I compare with undertow `accessLogType`, there is just one level configuration. Isn't that enough for ejb?
```
    <xs:complexType name="accessLogType">
        <xs:attribute name="pattern" use="optional" type="xs:string" default="common"/>
        <xs:attribute name="worker" use="optional" type="xs:string" default="default"/>
        <xs:attribute name="directory" use="optional" type="xs:string" default="${jboss.server.log.dir}"/>
        <xs:attribute name="relative-to" use="optional" type="xs:string" />
        <xs:attribute name="prefix" use="optional" type="xs:string" default="access_log."/>
        <xs:attribute name="suffix" use="optional" type="xs:string" default="log"/>
        <xs:attribute name="rotate" use="optional" type="xs:string" default="true"/>
        <xs:attribute name="use-server-log" use="optional" type="xs:string" default="false"/>
        <xs:attribute name="extended" use="optional" type="xs:string" default="false" />
        <xs:attribute name="predicate" use="optional" type="xs:string" />
    </xs:complexType>
```

Reply:
Unfortunately there's not much we can leverage from the logging subsystem as the output would likely need to be in a different format.

I do share the concerns though that we're going to end up with 4 ways of using different file handlers, console handlers, syslog handlers, etc.

Reply:
If we are going to duplicate existing configuration and implementation, then that sounds to me as proper fit for using capabilities [1] Different output can be managed by exposing configuration of capability.

What do you think?

[1] https://docs.jboss.org/author/display/WFLY/Working+with+WildFly+Capabilities

Reply:
Unfortunately I don't think that will work in this case. The reason being is the logging subsystem would be likely providing the base capabilities. This means it would need to be present in order for access logging to work. While I would guess most people do not remove the logging subsystem, a change like this would prevent it from being possible if they want to use EJB access logging.

I'm definitely up for suggestions too as I do share the concerns with essentially duplicate resources.

The other catch is going through logging likely won't have the information needed. You'd need to use a `java.util.logging.Formatter` to use any log handlers and the `java.util.logging.LogRecord` likely won't have all the information an EJB access event would need.

We do have an [event logger](https://github.com/wildfly/wildfly-core/tree/master/event-logger) project within WildFly Core that we could however consider extending to get a more standard output from. I would not be opposed to this, but we'd need to open a discussion about that.

Reply:
@chengfang Still comparing against undertow/src/main/resources/schema/wildfly-undertow_10_0.xsd .... would it be possible to maintain, at least, the closest XSD possible in order to minimize differences from the user point of view?